### ISSUE TITLE
Mainline android support

### ISSUE THREAD

**Issue by loonycyborg** [wesnoth/wesnoth#7477]
### Describe the desired feature

Currently wesnoth has third-party android port which goes up to version 1.14. Its page is here:
https://www.alessandropira.org/bfw_android/get.html
However it seems that author is inactive atm and there's still no 1.16 port. We should sooner or later get mainline support for building against android anyway. I see two options how to proceed from here:
- try to merge the linked third-party port. Which would be rather hard for following reasons: seems it's not actually maintained in a vcs repo but changes only available as separate archives. Also it has some extra features like partial install with ability to download campaigns/translations separately. Not even sure from where.
- Make our own port from scratch. There's good chance that there were advances in the ways SDL apps are ported to android since third party port was made. Also it would be easier to plug in our existing packaging framework.

There are other quality issues related to android. For example, preferably we should have native touch input support. Not sure how close we are in that regard.

**Comment by gfgtdf** [comment]
Do we currently have an idea how the touch input on android should look like in the end? I mean on desktop the usual input is still reliant on right-click-menus and hotkeys

**Comment by Pentarctagon** [comment]
@singalen How is touch handled on the iOS port?

**Comment by gfgtdf** [comment]
When exactly does it depend on hovers, i think having to to long-touches for every action might become annoying in the long run.

**Comment by soliton-** [comment]
https://github.com/singalen/wesnoth/issues/44

**Comment by CelticMinstrel** [comment]
Another option for the context menu is two-finger tap. It could also be a setting.

**Comment by loonycyborg** [comment]
We could use gestures too, but not sure how that would affect compatibility.

**Comment by Pentarctagon** [comment]
Though I'm guessing that the biggest issue will be getting the UI to work with a smartphone's form factor without something equivalent to the old tinygui. Honestly I'm curious if just rebuilding wesnoth as-is would even be able to open the titlescreen vs having a "not enough space" layout error.

**Comment by gfgtdf** [comment]
What was tinygui agan? I mea how exactly was it different ?

**Comment by Pentarctagon** [comment]
It was for support for very small resolution screens.

**Comment by gfgtdf** [comment]
Yes, but how exactly was it different and in particular what was bad about it ?

**Comment by Pentarctagon** [comment]
It looks like the last stable branch it was in was 1.8, so it's before when I was involved in development. My understanding though is that there wasn't anything wrong with it per se, it was just non-trivial to continue supporting while being barely used.

**Comment by CelticMinstrel** [comment]
Both the GUI2 system and ThemeWML are well-equipped for adjusting the layout based on resolution. If you wanted, you could easily build a _dramatically_ different titlescreen for a ridiculously small resolution, like 640x480 or even smaller. We already make use of this in a few screens, such as the add-on manager and the multiplayer create screen.

**Comment by Pentarctagon** [comment]
Right, it's certainly possible. I doubt it would be *easy* though - we've had a variety of issues where something breaks specifically on 800x600 due to the smaller size.

**Comment by CelticMinstrel** [comment]
I mean, it would be about as hard as designing an interface from scratch, I guess. Maybe a little harder given that it's limited to a very small size. Would probably also require a small about of C++ coding because the C++ is likely assuming things to exist that wouldn't be guaranteed to exist in the smaller version (they'd be in the widget hierarchy, but not always be visible, so not found by `find_widget`).

**Comment by alexpira** [comment]
Hi I am the owner of the quoted page and the developer of the third-party android port you are talking about.
As you said I am really not active on the port due to the lack of time and commitment, but if I can be of some help please let me know.
Biggest problem for 1.16 upgrade is that it would need an SDL upgrade. Keep in mind that every dependency must be built successfully on Android with NDK: pango, glib, etc... Upgrading a library is very time consuming and also quite complex.
With 1.14 I started basically from scratch with SDL, but now it needs to be done again, if you want to upgrade.

**Comment by alexpira** [comment]
@Pentarctagon ok... in theory :)
In practice it was not just a build tool problem: some libraries had functionalities which was not developed with Android in mind and almost nothing built ok on the first run.
But hey, time passed and maybe things now are easier...

**Comment by Pentarctagon** [comment]
Yeah, theory and practice are definitely different :P

It was used for a somewhat similar purpose already though, so I figured I'd mention it here too.

**Comment by loonycyborg** [comment]
> ok... in theory :) In practice it was not just a build tool problem: some libraries had functionalities which was not developed with Android in mind and almost nothing built ok on the first run. But hey, time passed and maybe things now are easier...

I managed to build all deps with only minor buildsystem fixes. But I didn't get around to building an apk yet. Seems it's mostly about using SDL's android template. Can you detail which exact functionality that is specific to android point you ended up implementing?

**Comment by alexpira** [comment]
@loonycyborg I basically rewrote the whole wesnoth-SDL graphics & input bridge. That allowed me to add an overlay layer (with on-screen buttons for special actions) and to manage in java all the input parsing for multi-finger actions.

I uploaded for you the changes I made to wesnoth sources for 1.14.17 here:

https://www.alessandropira.org/bfw_android/patch_wesnoth-1.14.17.zip

In SDL, to sum up all my changes you can start from the library package:

https://sourceforge.net/projects/wesnoth-on-android/files/1.14.9/Libraries-20191103.tar.gz

and check for the jni/SDL2-2.0.5/src/android_ap folder.

Hope this helps.

**Comment by loonycyborg** [comment]
> and check for the jni/SDL2-2.0.5/src/android_ap folder.

Everything is in C? Seems SDL2 at least provides a project template with lot of java code.

**Comment by Elvish-Hunter** [comment]
Another thing that might be interesting is that apparently later Android releases have become even stricter about accessing app files: a user on the subreddit had to use both a specific file manager and Google Drive to export a map made with our editor. Details here: https://new.reddit.com/r/wesnoth/comments/12c3qgy/how_to_export_custom_map_on_android/.

**Comment by alexpira** [comment]
> Everything is in C? Seems SDL2 at least provides a project template with lot of java code. Which exact things from there would need porting effort for SDL2? Maybe some got obsoleted by something SDL2 offers by itself?

In my port, java is responsible for starting up the game (this includes downloading wesnoth data and preparing the game folder, loading shared object, managing port specific options, creating the view with the main surface and kick-starting the main function) and manage touch events & audio.

What is missing in vanilla SDL2, which includes some android code, and I believe should be ported is:
- game data downloading and unpacking mechanism
- configurable on-screen keyboard which in my port includes keys for:
  - popping up and hide the software keyboard
  - in-game actions like "next unit", "show enemy move range",  "end turn", etc...
- multi finger wesnoth specific actions, which allows for example scrolling the map by dragging two fingers on the screen, defining gestures for the left & right mouse button, choosing between "pointer-follows-finger" mode (which is best for bigger devices like tablets) and relative mouse movement (which works better on smaller phones)

These are the main requirements. Of course, you'll probably stumble on a lot of minor things that needs to be changed like for example loading all the shared objects in the right order, not just SDL :)

**Comment by Zireael07** [comment]
@Elvish-Hunter Yep! Godot Android version also has run into similar trouble with accessing files

**Comment by alexpira** [comment]
> What is missing in vanilla SDL2, which includes some android code, and I believe should be ported is:

Also, something for access or at least import/export saved games is needed as the folder might not be accessible to the user.

On android each app has a private folder in the filesystem which on non-rooted phones is not accessible to the user. Alternatively, the app can ask for user permission for accessing the /sdcard folder, which is a shared space accessible to the user. The user can decide to grant and revoke this permission anytime, so if game data is installed there, every time the game starts the request for access must be reiterated if needed.

Another solution (which is not what I made in my port) can be to keep game data files in the private space and store saved games somewhere inside /sdcard.

**Comment by loonycyborg** [comment]
>     * game data downloading and unpacking mechanism
> 
Downloading from where? Currently we only officially support full game tarball releases and their builds for some platforms. I suppose piecewise downloading is nice but modern phones have so much storage that many won't notice wesnoth's 600-800Mb.

>     * configurable on-screen keyboard which in my port includes keys for:
>       
>       * popping up and hide the software keyboard
>       * in-game actions like "next unit", "show enemy move range",  "end turn", etc...
>
Ideally for upstream port those things should be made native parts of wesnoth's UI. For example auto-poping screen keyboard when text input elements are focused.
 
>     * multi finger wesnoth specific actions, which allows for example scrolling the map by dragging two fingers on the screen, defining gestures for the left & right mouse button, choosing between "pointer-follows-finger" mode (which is best for bigger devices like tablets) and relative mouse movement (which works better on smaller phones)
> 
Maybe we could provide enough specific gestures to avoid having a pointer in the first place.

**Comment by Pentarctagon** [comment]
> Downloading from where? Currently we only officially support full game tarball releases and their builds for some platforms. I suppose piecewise downloading is nice but modern phones have so much storage that many won't notice wesnoth's 600-800Mb.

529 MBs for the Linux 1.17 steam build, for reference.

**Comment by alexpira** [comment]
> Downloading from where? Currently we only officially support full game tarball releases and their builds for some platforms. I suppose piecewise downloading is nice but modern phones have so much storage that many won't notice wesnoth's 600-800Mb.

Maximum APK file size allowed on Google Play is much smaller (100Mb). There are mechanism to add an external pack but it requires code to unpack it anyway. Also, I don't know if that's supported on other repositories like F-Droid.

https://developer.android.com/google/play/expansion-files

**Comment by loonycyborg** [comment]
Potentially could add binaries to main apk and data to expansion files. Already doing something like that for steam. At least google play would be hosting them in this case.

**Comment by babaissarkar** [comment]
Completed by #7707 which was manually merged.
